### PR TITLE
MM-10188 extra large file upload

### PR DESCRIPTION
**Pull request by levb**
#### Summary
Support extra large file uploads (1Tb+)

- Implemented the change to bypass buffering when there are no applicable plugins
- Implemented utils.LargeBuffer
- Switched app.UploadFileX to use LargeBuffer
- Updated tests

Current performance improvements:
```
$ go test -v --run=nothing --bench=UploadFile -benchmem ./app
BenchmarkUploadFile/random-5Mb-gif-DoUploadFile_raw-8         	      20	  62801004 ns/op	21211826 B/op	    1009 allocs/op
BenchmarkUploadFile/random-5Mb-gif-UploadFileX_raw_chunked-8  	     200	   8943735 ns/op	 2140430 B/op	     136 allocs/op
BenchmarkUploadFile/random-5Mb-gif-UploadFiles-8              	       3	 522987944 ns/op	81938776 B/op	 3705029 allocs/op
BenchmarkUploadFile/random-5Mb-gif-UploadFileX_chunked-8      	       3	 458466725 ns/op	67183885 B/op	 3704449 allocs/op
BenchmarkUploadFile/random-2Mb-jpg-DoUploadFile_raw-8         	     200	   6672348 ns/op	 6032712 B/op	    8052 allocs/op
BenchmarkUploadFile/random-2Mb-jpg-UploadFileX_raw_chunked-8  	     300	   6103151 ns/op	 2140423 B/op	     135 allocs/op
BenchmarkUploadFile/random-2Mb-jpg-UploadFiles-8              	       3	 486272277 ns/op	70458458 B/op	 3718874 allocs/op
BenchmarkUploadFile/random-2Mb-jpg-UploadFileX_chunked-8      	       2	 515760044 ns/op	58288376 B/op	 3710983 allocs/op
BenchmarkUploadFile/zero-1024Mb-DoUploadFile_raw-8            	       1	3540935025 ns/op	6478521056 B/op	  262329 allocs/op
BenchmarkUploadFile/zero-1024Mb-UploadFileX_raw_chunked-8     	       2	 637574425 ns/op	 2141284 B/op	     142 allocs/op
BenchmarkUploadFile/zero-1024Mb-UploadFiles-8                 	       1	3084544538 ns/op	6478520888 B/op	  262329 allocs/op
BenchmarkUploadFile/zero-1024Mb-UploadFileX_chunked-8         	       2	 636665350 ns/op	 2140964 B/op	     137 allocs/op
```

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-10188
https://github.com/mattermost/mattermost-server/issues/8641

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [ ] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

STILL TODO:
- Add config for temp file encoding
- Update documentation with the new config settings

### CODE REVIEW

**Comment by levb** [comment]
@lieut-data @crspeller @jwilander I can use your opinions here, I think I hit a dead end for the time being. Once I try to run a test that invokes a FileWillBeUploaded plugin that streams a 200Mb file over the RPC-marshalled io.Reader I see it hang ~50% of the time while reading the input, and sometimes writing the output back to the server. From the reader/writer logs I could see that it was reading in 8Kb chunks, subsequently degrading to 1Kb chunks. I am not sure exactly how the RPC represents streams, but I'd assume that it preserves the `io` interface semantics and roundtrips for each function call.  If so, these are still very small chunks, and many O(E5) roundtrips involved.

I think we have 2 options here:
(a) keep digging into the RPC-based implementations (current or alternative, gRPC and what not) seeking either a way to tune io.Reader/io/Writer performance to an acceptable level. This is tempting but also can end up a time drain and not produce an acceptable outcome.

(b) acknowledge that the RPC mechanism is not suitable for large data transfer, and look for an alternative streaming architecture for processing such entities (http?). We can keep the current hook functional for images and other small files, but move to http in/out for larger files. Also, I don't understand, do we always call the local subprocess with the RPC, or (say, in Kubernetes) do we envision these calls load-balanced across instances?

Thoughts?

**Comment by lieut-data** [comment]
Personally, I'm leaning towards (b). I don't think it's reasonable to expect that we can stream every incoming file to every plugin regardless of size.

The main problem with (b) is the need to buffer the file on disk, which might violate some customer requirements. But I'm struggling to imagine a file-checking plugin that can do it's work entirely in memory anyway (unless it needed only metadata, for which we could add a new, earlier hook). So my proposal would be to reject file uploads above a set threshold /if/ there is a plugin registered to process the file contents.

**Comment by levb** [comment]
RE: ticket for benchmarking plugins and RPCs: coming up.

RE: @crspeller "would like to know why RPC isn't capable of this" I think the RPCs are _capable_ of this, but since they have to preserve the semantics of a function call on a stream, they probably make a roundtrip for each Read/Write. Given ~1.5Kb average read size and assuming a ~100ns latency/overhead per call we get *at most* 15Mb/s transfer rate, each way. Hardly an efficient way of moving bytes on the same machine.

Now, in my case the small Read size is probably coming from `io.Copy` into /dev/null, but we can't really control how plugins Read/Write their streams.

RE: Option (b). I too agree that it's the way to go, need to think through the specific options and benchmark.

**Comment by crspeller** [comment]
I understand you explanation for (a) now. Thanks!

**Comment by levb** [comment]
Please ignore updates for now, got it to work with temp files, but still need to polish and test a lot, will fully squash before I re-present for review.

**Comment by levb** [comment]
This is now ready for review

File: plugin/client_rpc.go [hunk 1]
```diff
@@ -365,81 +366,103 @@ func init() {
...
+
+func (wrapper *FileWillBeUploadedWrapper) Read(p []byte) (int, error) {
+	return 0, fmt.Errorf("Unreachable code. Fake FileWillBeUploadedWrapper.Read called.")
+}
+
+func (wrapper *FileWillBeUploadedWrapper) Write(p []byte) (int, error) {
+	return 0, fmt.Errorf("Unreachable code. Fake FileWillBeUploadedWrapper.Write called.")
+}
+
 type Z_FileWillBeUploadedArgs struct {
```
Comment (lieut-data) [inline plugin/client_rpc.go]: Is this a backwards incompatible change?

File: utils/large_buffer.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+}
+
+var ErrLargeBufferMustClose = errors.New("LargeBuffer must be Closed before reading from file")
+
+// TODO: consider using https://github.com/djherbis/buffer for better performance
+
+// Type LargeBuffer implements a buffer backed by a bytes.Buffer up to
+// maxMemoryBuffer bytes. It uses a temporary file for any extra payload bytes.
+// The file can be optionally encoded to alleviate concerns about saving
+// maliciopus contents to disk.
```
Comment (lieut-data) [inline utils/large_buffer.go]:
```suggestion
// malicious contents to disk.
```

File: api4/file_test.go [hunk 3]
```diff
@@ -315,6 +316,10 @@ func TestUploadFiles(t *testing.T) {
 	// Get better error messages
 	th.App.UpdateConfig(func(cfg *model.Config) { *cfg.ServiceSettings.EnableDeveloper = true })
 
+	// Lower the in-memory buffer size, shouldn't get in the way since no
+	// plugins are involved
+	th.App.UpdateConfig(func(cfg *model.Config) { *cfg.FileSettings.MaxMemoryBuffer = 8 * 1024 * 1024 })
```
[the remark this answers is not in the document]
Comment (levb) [reply api4/file_test.go]: I eliminated the need for config changes within the test, to keep it simpler. The only one that needed a different config was "uploads disabled", I moved it into a separate test/app object.

File: api4/file_test.go [hunk 4]
```diff
@@ -554,49 +559,25 @@ func TestUploadFiles(t *testing.T) {
...
-				}
-			},
-		},
-		// File too large (chunked, simple POST only, multipart would've been redundant with above)
-		{
-			title:                  "File too large chunked",
+			title:                  "Too large chunked",
+			names:                  []string{"50Mb-plus-stream"},
 			useChunkedInSimplePost: true,
 			skipMultipart:          true,
```
[the remark this answers is not in the document]
Comment (levb) [reply api4/file_test.go]: Done.

File: api4/file_test.go [hunk 5]
```diff
@@ -554,49 +559,25 @@ func TestUploadFiles(t *testing.T) {
...
-				maxFileSize := *a.Config().FileSettings.MaxFileSize
-				a.UpdateConfig(func(cfg *model.Config) { *cfg.FileSettings.MaxFileSize = 279590 })
-				return func(a *app.App) {
-					a.UpdateConfig(func(cfg *model.Config) { *cfg.FileSettings.MaxFileSize = maxFileSize })
-				}
-			},
 		},
 		{
-			title:                 "Error stream too large",
+			title:                 "Too large",
```
[the remark this answers is not in the document]
Comment (levb) [reply api4/file_test.go]: Per your suggestion I included all of the test flags to clarify

File: app/file.go [hunk 6]
```diff
@@ -576,105 +627,169 @@ func (a *App) UploadFileX(channelId, name string, input io.Reader,
...
+		if err != nil {
+			t.appErr = t.newAppError("api.file.upload_file.read_request.app_error",
+				err.Error(), http.StatusBadRequest)
+			return false
+		}
+		t.lbuf.Clear()
+		t.lbuf = newLbuf
+		t.memoryLimitedTeeInput = nil
+		t.sizeLimitedInput = nil
+		// TODO t.fileinfo.Size = int64(newLbuf.Len())
```
[the remark this answers is not in the document]
Comment (levb) [reply app/file.go]: Whew, was a good catch, thanks @lieut-data! Fixed now. The issue was that the Size is indeterminate until the file is written out. I still would like to add an explicit test for it, with a plugin.

File: app/file.go [hunk 7]
```diff
@@ -722,11 +835,22 @@ func (t *uploadFileTask) preprocessImage() *model.AppError {
...
+func (t *uploadFileTask) needProcessImage() bool {
+	return !t.Raw && t.fileinfo.IsImage() && !t.imageIsTooLarge
+}
+
+func (t *uploadFileTask) needDecodeImage() bool {
+	return t.needProcessImage() && t.decoded == nil
+}
+
+func (t *uploadFileTask) decodePipedImage(in io.Reader) {
+	// Since in is a pipe reader, need to fully drain it to avoid broken pipe or hangs.
```
[the remark this answers is not in the document]
Comment (levb) [reply app/file.go]: I clarified the code. It was correct, but indeed confusing.

File: plugin/client_rpc.go (hunk 1 above)
Comment (levb) [reply plugin/client_rpc.go]: Yes, I'll take the discussion into the main PR comment thread.

File: app/file.go [hunk 8]
```diff
@@ -565,6 +606,16 @@ func (a *App) UploadFileX(channelId, name string, input io.Reader,
 		o(t)
 	}
 	t.init(a)
+	defer t.lbuf.Clear()
+
+	// We may run into the entityTooLarge error in a Read() call, where there's no way to
```
[the remark this answers is not in the document]
Comment (levb) [reply app/file.go]: I refactored the relevant code for clarity, but left the `defer` there. Unfortunately, I do not see how I can reliably return a known error. However I got rid of t.appErr and made the use-case here explicit, a file too large.

File: app/file.go [hunk 9]
```diff
@@ -576,105 +627,169 @@ func (a *App) UploadFileX(channelId, name string, input io.Reader,
...
+			writeError = result.Err
+			return
+		}
+	}()
+
+	if t.needProcessImage() {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			if t.needDecodeImage() {
```
[the remark this answers is not in the document]
Comment (levb) [reply app/file.go]: Clarified.

**Comment by levb** [comment]
@lieut-data Thanks for the thorough review and catching a couple materially-significant issues!
I refactored to clarify the confusing parts that you pointed out. I think I addressed your specific comments.

The big outstanding issues are:
1. This is currently implemented as a backwards incompatible change in the _marshalled_ version of FileWillBeUploaded, breaking compatibility with the installed _binary_ plugins. They'd need to be recompiled and reinstalled API change.
2. Compliance/security issues of buffering to disk.
3. Passing around a local file name is not good for Kubernetes/distributed arch.

Going in reverse,
(3) can be addressed by sending around a (temporary) URL that can be served by MM server. 
(2) can be addressed by setting `MaxMemoryBuffer` > `MaxFileSize` in the config

The binary compatibility issue is a big one. I can break this into a new hook, say `LargeFileWillBeUploaded` or keep the old functionality/and structs  intact but extend them with the new fields. Not crazy about either. Maybe once we have all other implementation issues out of the way, we could do a quick discussion with @crspeller and @jwilander of how to best handle backwards binary plugin compatibility.

File: utils/large_buffer_test.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+			for _, chunk := range tc.chunks {
+				n, err := lbuf.Write(chunk)
+				a.Empty(err, "Expected no error")
+				a.Equal(len(chunk), n, "Wrong n returned (bytes written)")
+			}
+
+			a.Equal(lbuf.buffer.Bytes(), tc.expectedBufferedBytes, "Buffered contents differs")
+
+			err := lbuf.Close()
+			a.Empty(err, "Expected no error")
```
Comment (hanzei) [inline utils/large_buffer_test.go]: Nit: Why not use `a.Nil()` and be explicit what we expect?

File: utils/large_buffer_test.go [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+			expectedFileBytes:     []byte("=BSf"),
+			expectedReadBackBytes: []byte("0123456789ABCDEFXYZ"),
+		},
+	} {
+		testfunc := func(t *testing.T, lbuf *LargeBuffer) {
+			a := assert.New(t)
+			for _, chunk := range tc.chunks {
+				n, err := lbuf.Write(chunk)
+				a.Empty(err, "Expected no error")
+				a.Equal(len(chunk), n, "Wrong n returned (bytes written)")
```
Comment (hanzei) [inline utils/large_buffer_test.go]: Nit: You can use `a.Len()` https://godoc.org/github.com/stretchr/testify/assert#Len here.

**Comment by lieut-data** [comment]
Sorry for the delay on this, @levb -- GitHub removed the review from my list, and it completely slipped off my radar :(.

I've added myself back to give it a pass soon!

File: utils/large_buffer_test.go (hunk 10 above)
Comment (lieut-data) [reply utils/large_buffer_test.go]: Or better yet, `a.NoError()`?

**Comment by hanzei** [comment]
Frezzing this one for now until we/Lev decided if/how we want this.

**Comment by lieut-data** [comment]
@levb, is the plan to still keep this PR open to eventually circle back, or would it make sense to start with a fresh PR at that point?

**Comment by levb** [comment]
I want to keep this one, but need to get to it for a few solid days to figure out (a) RPC compatibility issues, and (b) maybe serving files over localhost or domain socket http rather than (RPC - before, or sending filenames around)

**Comment by jasonblais** [comment]
@levb curious on your thoughts whether we should keep this PR open or close for now and reference it in the related Jira ticket? https://mattermost.atlassian.net/browse/MM-10188

**Pull request closed without merging**